### PR TITLE
Support for south SoNic added

### PR DESCRIPTION
**Pull request by cshivashgit**
Support for south SoNic daemon added

### CODE REVIEW

File: src/south/sonic-interface/main.cpp [hunk 1]
```diff
@@ -1,129 +1,463 @@
...
+{
+  std::stringstream url;
+  libyang::S_Context ly_ctx = session->get_context();
+
+  CURL *curl;
+  CURLcode res;
+  struct curl_fetch_st curl_fetch;
+  struct curl_fetch_st *fetch = &curl_fetch;
+  fetch->payload = (char *) calloc (1, sizeof(fetch->payload));
+  char *payload = (char *) calloc (1, MAX_PAYLOAD_SIZE);
```
Comment (ishidawataru) [inline src/south/sonic-interface/main.cpp:223]: why needed?

Comment (ishidawataru) [reply src/south/sonic-interface/main.cpp:223]: We may be able to remove this as well

https://github.com/microsonic/goldstone-mgmt/pull/23/files#diff-f5a830eeff1bf48f8cc57fdebfd8cd82R20

File: src/south/sonic-interface/main.cpp [hunk 2]
```diff
@@ -1,129 +1,463 @@
...
+  p->payload = (char *) realloc (p->payload, p->size + realsize + 1);
+
+  /* check buffer */
+  if (p->payload == NULL)
+    {
+      fprintf (stderr, "ERROR: Failed to expand buffer in curl_callback");
+      /* free buffer */
+      free (p->payload);
+      /* return */
+      return 1;
```
Comment (ishidawataru) [inline src/south/sonic-interface/main.cpp:181]: is this the correct return value when curl is failed?

File: src/south/sonic-interface/main.cpp [hunk 3]
```diff
@@ -1,129 +1,463 @@
...
+  /* check payload */
+  if (fetch->payload == NULL)
+    {
+      /* log error */
+      fprintf (stderr, "ERROR: Failed to allocate payload in curl_fetch_url");
+      /* return error */
+      return 0;
+    }
+  fetch->size = 0;
+  url << "https://" << intfIp << "/restconf/data";
```
Comment (ishidawataru) [inline src/south/sonic-interface/main.cpp:234]: please support `http://` as well.

File: src/south/sonic-interface/main.cpp [hunk 4]
```diff
@@ -1,129 +1,463 @@
...
+                                 libyang::S_Data_Node &parent,
+                                 void *private_data)
+{
+  std::stringstream url;
+  libyang::S_Context ly_ctx = session->get_context();
+
+  CURL *curl;
+  CURLcode res;
+  struct curl_fetch_st curl_fetch;
+  struct curl_fetch_st *fetch = &curl_fetch;
```
Comment (ishidawataru) [inline src/south/sonic-interface/main.cpp:221]: no `struct` needed in C++

File: src/south/sonic-interface/main.cpp [hunk 5]
```diff
@@ -1,129 +1,463 @@
...
+{
+  if (argc != 2)
+    {
+      std::cerr << "Usages: ip_address" << std::endl;
+      exit (0);
+    }
 
-    auto controller = SonicController(sess);
+  /* Management IP address to be passed */
+  intfIp = argv[1];
```
Comment (ishidawataru) [inline src/south/sonic-interface/main.cpp:450]: Please use `getopt_long` like TAI south

File: Makefile [hunk 6]
```diff
@@ -96,6 +96,9 @@ init:
 	sysrepoctl -s $(GS_YANG_REPO) --install $(GS_YANG_REPO)/goldstone-onlp.yang
 	sysrepoctl -s $(GS_YANG_REPO) --install $(GS_YANG_REPO)/goldstone-tai.yang
 	sysrepoctl -s $(GS_YANG_REPO) --install $(GS_YANG_REPO)/goldstone-sonic-interface.yang
+	sysrepoctl -s $(GS_YANG_REPO) --install $(GS_YANG_REPO)/sonic-common.yang
+	sysrepoctl -s $(GS_YANG_REPO) --install $(GS_YANG_REPO)/sonic-port.yang
+	sysrepoctl -s $(GS_YANG_REPO) --install $(GS_YANG_REPO)/sonic-interface.yang
```
Comment (ishidawataru) [inline Makefile:101]: Can we add the right submodule which contains these SONiC YANG models and point them directly like what we're doing for OpenConfig models?

File: src/south/sonic-interface/main.cpp [hunk 7]
```diff
@@ -1,129 +1,463 @@
...
+#include <stdlib.h>
+#include <unistd.h>
+#include <string.h>
+#include <curl/curl.h>
 
-//#include <swss-common/redispipeline.h>
-//#include <swss-common/dbconnector.h>
+#define MAX_PAYLOAD_SIZE 10000
 
+char *intfIp;
```
[the remark this answers is not in the document]
Comment (ishidawataru) [reply src/south/sonic-interface/main.cpp:22]: we can store the URL of the MGMT server ( e.g. "http://127.0.0.1" ) then the user can specify `http` / `https` or port if needed

File: src/south/sonic-interface/main.cpp (hunk 3 above)
Comment (cshivashgit) [reply src/south/sonic-interface/main.cpp]: Done

File: src/south/sonic-interface/main.cpp (hunk 5 above)
Comment (cshivashgit) [reply src/south/sonic-interface/main.cpp]: Done

File: Makefile (hunk 6 above)
Comment (cshivashgit) [reply Makefile]: Done

File: src/south/sonic-interface/main.cpp (hunk 1 above)
Comment (cshivashgit) [reply src/south/sonic-interface/main.cpp]: Done, Removed payload

File: src/south/sonic-interface/main.cpp [hunk 8]
```diff
@@ -1,129 +1,463 @@
...
+* Function Name : get_index_of_yang                                          *
+* Description   : To get the primary key of yang data                        *
+* Input         : libyang::S_Data_Node &parent                               *
+* Output        : char                                                       *
+*****************************************************************************/
+char
+*get_index_of_yang (libyang::S_Data_Node &parent)
+{
+  struct lyd_node *lyd = parent->C_lyd_node();
+  struct lys_node *lys = lyd->schema;
```
[the remark this answers is not in the document]
Comment (cshivashgit) [reply src/south/sonic-interface/main.cpp]: Done

File: src/south/sonic-interface/main.cpp (hunk 7 above)
[the remark this answers is not in the document]
Comment (cshivashgit) [reply src/south/sonic-interface/main.cpp]: Done

**Comment by ishidawataru** [comment]
@cshivashgit Thanks for the update. Squashed and merged

**Pull request merged**